src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp

```diff
@@ -0,0 +1,554 @@
+
+  QString lastError;
+  QgsPointCloudLayerProfileResults::PointResult *pointData = mResults->results.data();
+  const int size = mResults->results.size();
+  for ( int i = 0; i < size; ++i, ++pointData )
+  {
+    if ( mFeedback->isCanceled() )
+      return false;
+
+    pointData->distance = startDistanceOffset + originalCurveGeos.lineLocatePoint( pointData->x, pointData->y, &lastError );
+    pointData->curveDistance = originalCurveGeos.distance( pointData->x, pointData->y );
+
+    mResults->minZ = std::min( pointData->z, mResults->minZ );
+    mResults->maxZ = std::max( pointData->z, mResults->maxZ );
+  }
+
+  return true;
+}
+
+QgsAbstractProfileResults *QgsPointCloudLayerProfileGenerator::takeResults()
+{
+  return mResults.release();
+}
+
+QgsFeedback *QgsPointCloudLayerProfileGenerator::feedback() const
+{
+  return mFeedback.get();
+}
+
+QVector<IndexedPointCloudNode> QgsPointCloudLayerProfileGenerator::traverseTree( const QgsPointCloudIndex *pc, IndexedPointCloudNode n, double maxErrorPixels, double nodeErrorPixels, const QgsDoubleRange &zRange )
+{
+  QVector<IndexedPointCloudNode> nodes;
+
+  if ( mFeedback->isCanceled() )
+  {
+    return nodes;
+  }
+
+  const QgsDoubleRange nodeZRange = pc->nodeZRange( n );
+  const QgsDoubleRange adjustedNodeZRange = QgsDoubleRange( nodeZRange.lower() * mZScale + mZOffset, nodeZRange.upper() * mZScale + mZOffset );
+  if ( !zRange.isInfinite() && !zRange.overlaps( adjustedNodeZRange ) )
+    return nodes;
+
+  const QgsRectangle nodeMapExtent = pc->nodeMapExtent( n );
+  if ( !mMaxSearchExtentInLayerCrs.intersects( nodeMapExtent ) )
+    return nodes;
+
+  const QgsGeometry nodeMapGeometry = QgsGeometry::fromRect( nodeMapExtent );
+  if ( !mSearchGeometryInLayerCrsGeometryEngine->intersects( nodeMapGeometry.constGet() ) )
+    return nodes;
+
+  nodes.append( n );
+
+  double childrenErrorPixels = nodeErrorPixels / 2.0;
+  if ( childrenErrorPixels < maxErrorPixels )
+    return nodes;
+
+  const QList<IndexedPointCloudNode> children = pc->nodeChildren( n );
+  for ( const IndexedPointCloudNode &nn : children )
+  {
+    nodes += traverseTree( pc, nn, maxErrorPixels, childrenErrorPixels, zRange );
+  }
+
+  return nodes;
+}
+
+int QgsPointCloudLayerProfileGenerator::visitNodesSync( const QVector<IndexedPointCloudNode> &nodes, QgsPointCloudIndex *pc, QgsPointCloudRequest &request )
+{
+  int nodesDrawn = 0;
+  for ( const IndexedPointCloudNode &n : nodes )
+  {
+    if ( mFeedback->isCanceled() )
+      break;
+
+    std::unique_ptr<QgsPointCloudBlock> block( pc->nodeData( n, request ) );
+
+    if ( !block )
+      continue;
+
+    visitBlock( block.get() );
+
+    ++nodesDrawn;
+  }
+  return nodesDrawn;
+}
+
+int QgsPointCloudLayerProfileGenerator::visitNodesAsync( const QVector<IndexedPointCloudNode> &nodes, QgsPointCloudIndex *pc, QgsPointCloudRequest &request )
+{
+  int nodesDrawn = 0;
+
+  // see notes about this logic in QgsPointCloudLayerRenderer::renderNodesAsync
```

Comment:
You have probably based this on an earlier version of the code, recently I have simplified the logic (and made it more efficient) in QgsPointCloudLayerRenderer::renderNodesAsync() - see #48211

Reply:
Thanks, fixed